src/bot.ts

```diff
@@ -552,36 +552,63 @@ async function withRetries<T>(
     task: () => Promise<T>,
     signal?: AbortSignal,
 ): Promise<T> {
+    // Set up delays between retries
+    const INITIAL_DELAY = 50; // ms
+    let lastDelay = INITIAL_DELAY;
+
+    // Define error handler
+    /**
+     * Determines the error handling strategy based on various error types.
+     * Sleeps if necessary, and returns whether to retry or rethrow an error.
+     */
+    async function handleError(error: unknown) {
+        let delay = false;
+        let strategy: "retry" | "rethrow" = "rethrow";
+
+        if (error instanceof HttpError) {
+            delay = true;
+            strategy = "retry";
+        } else if (error instanceof GrammyError) {
+            if (error.error_code >= 500) {
+                delay = true;
+                strategy = "retry";
+            } else if (error.error_code === 429) {
+                const retryAfter = error.parameters.retry_after;
+                if (retryAfter !== undefined) {
+                    // sleep directly, don't let this affect the backoff
+                    await sleep(retryAfter, signal);
+                } else {
+                    delay = true;
+                }
+                strategy = "retry";
+            }
+        }
+
+        if (delay) {
+            // Do not sleep for the first retry
+            if (lastDelay !== INITIAL_DELAY) {
+                await sleep(lastDelay, signal);
+            }
+            const oneHour = 1 * 60 * 60 * 1000; // ms
+            lastDelay = Math.min(oneHour, 2 * lastDelay);
```

Comment:
Maybe add a randomised part to that number like ±1%? This prevents that all failing grammY clients with the same logic are hitting roughly at the same time again. If the server is under heavy load or whatever the reason was, its not a good idea to come back all together next time again 👻

Reply:
I'm not quite sure I understand how this could happen. There's nothing global that syncs up the clients, so the requested should technically already be spread out, no?

Reply:
Assuming there are 10 bots running hundreds of requests. Then the server fails due to whatever reason. All of the bots start the backdown. As the server failed synchronously for every bot (its the same server) all bots are synced up. At some point the server recovers and is available again. Then it gets hit by all the bots with the same back up delay logic at nearly the same time. Which might cause an overload, kill the server, fill up queues, …

Reply:
This code is only used during startup. That means you'll have to launch all of these 10 bots exactly at the same time.

Reply:
ah right, we are at the bot.init. Then yeah, its fine.